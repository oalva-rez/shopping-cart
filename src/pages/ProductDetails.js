import { React, useState } from "react";
import { useParams } from "react-router-dom";

export default function ProductDetails(props) {
  const { id } = useParams();
  const [product, setProduct] = useState(() =>
    props.products.filter((prod) => Number(prod.id) === Number(id))
  );
  const [quantity, setQuantity] = useState(1);
  return (
    <div className="content">
      <div className="product">
        <div className="image">
          <img src={product[0].image} alt="" />
        </div>
        <div className="details">
          <h2>{product[0].title}</h2>
          <h3>${product[0].price}</h3>
          <p>{product[0].description}</p>
          <p>
            Lorem ipsum dolor sit amet consectetur, adipisicing elit. Animi
            voluptate neque harum. Quam facere accusamus exercitationem in
            quidem mollitia eligendi porro eos voluptates iure incidunt,
            laudantium sed harum omnis quasi?
          </p>
          <div className="qty">
            <div className="qty">
              <button
                className="qty-count qty-count--minus"
                data-action="minus"
                onClick={() => {
                  setQuantity((prev) => (prev === 1 ? 1 : prev - 1));
                }}
              >
                -
              </button>
              <input
                className="product-qty"
                type="number"
                name="product-qty"
                min="0"
                max="10"
                value={quantity}
                readOnly
              />
              <button
                className="qty-count qty-count--add"
                data-action="add"
                onClick={() => {
                  setQuantity((prev) => prev + 1);
                }}
              >
                +
              </button>
            </div>
            <button
              onClick={() => {
                props.updateCart(product[0], quantity);
              }}
            >
              Add to cart
            </button>
          </div>
        </div>
      </div>
    </div>
  );
}
