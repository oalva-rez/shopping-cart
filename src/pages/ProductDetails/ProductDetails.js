import { React, useState } from "react";
import { useParams } from "react-router-dom";

export default function ProductDetails(props) {
  const { id } = useParams();
  const [product, setProduct] = useState(() =>
    props.products.filter((prod) => Number(prod.id) === Number(id))
  );
  const [quantity, setQuantity] = useState(1);

  return (
    <div className="content">
      <div className="product-main">
        <div className="image">
          <img src={product[0].image} alt="" />
        </div>
        <div className="details">
          <h2>{product[0].title}</h2>
          <h3>${product[0].price}</h3>
          <p>{product[0].description}</p>
          <p>
            Lorem ipsum dolor sit amet consectetur, adipisicing elit. Animi
            voluptate neque harum. Quam facere accusamus exercitationem in
            quidem mollitia eligendi porro eos voluptates iure incidunt,
            laudantium sed harum omnis quasi?
          </p>
          <div className="qty-container">
            <div className="qty">
              <button
                className="qty-count qty-count--minus"
                data-action="minus"
                onClick={() => {
                  setQuantity((prev) => (prev === 1 ? 1 : prev - 1));
                }}
              >
                -
              </button>
              <div className="qty-num">{quantity}</div>
              <button
                className="qty-count qty-count--add"
                data-action="add"
                onClick={() => {
                  setQuantity((prev) => prev + 1);
                }}
              >
                +
              </button>
            </div>
            <button
              className="add-to-cart"
              onClick={() => {
                props.updateCart(product[0], quantity);
              }}
            >
              Add to cart
            </button>
          </div>
        </div>
      </div>
    </div>
  );
}
