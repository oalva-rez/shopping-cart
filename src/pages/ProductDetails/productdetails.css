.product-main {
  display: flex;
}
.product-main .image {
  display: grid;
  place-content: center;
  padding: 40px 0;
}
.product-main img {
  width: 400px;
  height: 600px;
}
.product-main > * {
  flex: 1;
}

.details {
  display: flex;
  flex-direction: column;
  justify-content: center;
}
.details p {
  width: 70%;
}

.qty-container {
  margin-top: 20px;
}

.qty button {
  width: 30px;
  height: 30px;
  background-color: #fff;
  border: 1px solid #000;
  font-size: 20px;
  font-weight: 600;
  cursor: pointer;
}
.qty {
  display: flex;
}
.qty .qty-num {
  text-align: center;
  display: grid;
  place-content: center;
  width: 35px;
  font-size: 1.2rem;
  font-weight: 500;
}

.add-to-cart {
  width: 200px;
  height: 30px;
  font-weight: 500;
  margin-top: 10px;
  font-size: 1.1rem;
  background-color: black;
  color: white;
  border: none;
  cursor: pointer;
}
