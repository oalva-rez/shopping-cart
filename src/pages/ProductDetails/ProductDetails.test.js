import { render, screen } from "@testing-library/react";
import ProductDetails from "./ProductDetails";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import { BrowserRouter } from "react-router-dom";

const mockUpdateCart = jest.fn();
const mockProducts = [
  {
    description: "Product description",
    id: 1,
    image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
    price: 10,
    title: "Product 1",
  },
  {
    description: "Product 2 description",
    id: 2,
    image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
    price: 11,
    title: "Product 2",
  },
];
const mockProductDetails = () => {
  return (
    <BrowserRouter>
      <ProductDetails products={mockProducts} updateCart={mockUpdateCart} />
    </BrowserRouter>
  );
};

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    id: 1,
  }),
  useRouteMatch: () => ({
    url: "/products/1",
  }),
}));

describe("ProductDetails", () => {
  it("should match snapshot", () => {
    const { container } = render(mockProductDetails());
    expect(container).toMatchSnapshot();
  });
  it("should render product 1`s title", () => {
    render(mockProductDetails());
    expect(
      screen.getByRole("heading", { name: "Product 1" })
    ).toBeInTheDocument();
  });
  it('should change not change quantity when "-" button is clicked and is already 1', () => {
    render(mockProductDetails());
    const minusButton = screen.getByRole("button", { name: "-" });
    userEvent.click(minusButton);
    expect(screen.getByText("1", { selector: "div" })).toBeInTheDocument();
  });
  it('should add one to quantity when "+" button is clicked', () => {
    render(mockProductDetails());
    const plusButton = screen.getByRole("button", { name: "+" });
    userEvent.click(plusButton);
    expect(screen.getByText("2", { selector: "div" })).toBeInTheDocument();
  });
  it('should add product to cart when "Add to cart" button is clicked', () => {
    render(mockProductDetails());
    const addToCartButton = screen.getByRole("button", { name: "Add to cart" });
    userEvent.click(addToCartButton);
    expect(mockUpdateCart).toHaveBeenCalledTimes(1);
  });
});
