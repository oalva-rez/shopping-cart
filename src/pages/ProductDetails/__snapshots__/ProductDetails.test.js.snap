// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails should match snapshot 1`] = `
<div>
  <div
    class="content"
  >
    <div
      class="product-main"
    >
      <div
        class="image"
      >
        <img
          alt=""
          src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
        />
      </div>
      <div
        class="details"
      >
        <h2>
          Product 1
        </h2>
        <h3>
          $
          10
        </h3>
        <p>
          Product description
        </p>
        <p>
          Lorem ipsum dolor sit amet consectetur, adipisicing elit. Animi voluptate neque harum. Quam facere accusamus exercitationem in quidem mollitia eligendi porro eos voluptates iure incidunt, laudantium sed harum omnis quasi?
        </p>
        <div
          class="qty-container"
        >
          <div
            class="qty"
          >
            <button
              class="qty-count qty-count--minus"
              data-action="minus"
            >
              -
            </button>
            <div
              class="qty-num"
            >
              1
            </div>
            <button
              class="qty-count qty-count--add"
              data-action="add"
            >
              +
            </button>
          </div>
          <button
            class="add-to-cart"
          >
            Add to cart
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
